### PR TITLE
Adding code for ability to set metadata address in openid validation …

### PR DESCRIPTION
**Pull request by aderderian**
…options. Helpful for Azure B2C scenarios. This PR contains code for making Metadata (well known) address set via the token validation module. This will force a download of the information for there instead of trying to use the issuer/authority. Helpful in scenarios such as Azure B2C.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=12265) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=12265) before we can accept your contribution.<br/><br/>:x: aderderian [sign now](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=12265)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/OrchardCMS/OrchardCore?pullRequest=12265) it.</sub>

File: src/OrchardCore.Modules/OrchardCore.OpenId/Settings/OpenIdValidationSettings.cs [hunk 1]
```diff
@@ -8,5 +8,8 @@ public class OpenIdValidationSettings
         public Uri Authority { get; set; }
         public bool DisableTokenTypeValidation { get; set; }
         public string Tenant { get; set; }
+#nullable enable
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Settings/OpenIdValidationSettings.cs:11]: I'm all for using nullable references but I'm not sure what's the policy in place in OC. I'm not so sure it's worth decorating only select properties in a file.

/cc @sebastienros

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml [hunk 2]
```diff
@@ -28,6 +28,13 @@
     <span class="hint">@T["The address of the remote OpenID Connect server. Using a HTTPS address is strongly recommended."]</span>
 </div>
 
+<div class="mb-3" asp-validation-class-for="MetadataAddress">
+    <label asp-for="MetadataAddress">@T["Well Known Metadata Address"]</label>
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml:32]: `Well Known Metadata Address` may not be the best choice: the .well-known endpoints are not necessarily related to OIDC and you have tons of other standard endpoints for many other scenarios.

Maybe simply `Metadata Address`? or `OpenID Connect Metadata Address`?

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml [hunk 3]
```diff
@@ -28,6 +28,13 @@
...
+    <input asp-for="MetadataAddress" class="form-control" />
+    <span asp-validation-for="MetadataAddress"></span>
+    <span class="hint">@T["The address of the metatdata endpoint. Optional. (https://{your base url}/.well-known/openid-configuration)"]</span>
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml:35]: We need to be clearer about the fact it's only for advanced scenarios where the identity provider doesn't host it at the usual location and that `/.well-known/openid-configuration` will be used by default.

**Review by kevinchalet — CHANGES_REQUESTED** [review]
Thanks for your PR.

You'll need to add some validation code here: https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdValidationService.cs#L129

New rules we need:
  - We can't set `MetadataAddress` when a tenant is configured.
  - When set, it must be a well-formed URI (it can be an absolute or relative URL ).
  - When set, it can't include a query or fragment part.

We should also consider adding a HTTP/HTTPS requirement for both the authority/issuer and the metadata address: OpenIddict itself allows you to configure non-HTTP-based addresses as a non-standard extension but it requires custom code to take care of sending the request and it's highly unlikely we'll ever do that in Orchard.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml (hunk 2 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml:32]: We don't use them right now, we assume strings can be null.

**Comment by kevinchalet** [comment]
@MichaelPetrinolis good point!

For consistency, we may also want to add built-in support for `MetadataAddress` in the client too: the underlying stack is different, but it should work more or less the same way, except regarding how the absolute metadata address is computed if a relative URL is provided. OpenIddict uses a more complete logic that we may want to reuse: https://github.com/openiddict/openiddict-core/blob/dev/src/OpenIddict.Validation/OpenIddictValidationConfiguration.cs#L109-L134

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml (hunk 3 above)
Comment (aderderian) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml]: Updated

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml (hunk 2 above)
Comment (aderderian) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml]: Updated

File: src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdValidationService.cs [hunk 4]
```diff
@@ -110,15 +110,60 @@ public async Task<ImmutableArray<ValidationResult>> ValidateSettingsAsync(OpenId
                     }));
                 }
 
+                if (settings.Authority.Scheme != Uri.UriSchemeHttps)
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdValidationService.cs:113]: Please update this line to allow HTTP addresses and use a case-insensitive comparison.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdValidationService.cs [hunk 5]
```diff
@@ -110,15 +110,60 @@ public async Task<ImmutableArray<ValidationResult>> ValidateSettingsAsync(OpenId
...
+            {
+                if (!settings.MetadataAddress.IsAbsoluteUri || !settings.MetadataAddress.IsWellFormedOriginalString())
+                {
+                    results.Add(new ValidationResult(S["The specified metadata address is not valid."], new[]
+                    {
+                        nameof(settings.MetadataAddress)
+                    }));
+                }
+
+                if (settings.MetadataAddress.Scheme != Uri.UriSchemeHttps)
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdValidationService.cs:141]: Same remark here 😃

File: src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdValidationService.cs [hunk 6]
```diff
@@ -110,15 +110,60 @@ public async Task<ImmutableArray<ValidationResult>> ValidateSettingsAsync(OpenId
...
+                    {
+                        nameof(settings.MetadataAddress)
+                    }));
+                }
+
+                if (!String.IsNullOrEmpty(settings.Tenant))
+                {
+                    results.Add(new ValidationResult(S["No metatada address can be set when using another tenant."], new[]
+                    {
+                    nameof(settings.MetadataAddress)
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Services/OpenIdValidationService.cs:161]: Please fix the indentation here.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Settings/OpenIdValidationSettings.cs [hunk 7]
```diff
@@ -8,5 +8,7 @@ public class OpenIdValidationSettings
         public Uri Authority { get; set; }
         public bool DisableTokenTypeValidation { get; set; }
         public string Tenant { get; set; }
+        public Uri MetadataAddress { get; set; }
+
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Settings/OpenIdValidationSettings.cs:12]: Here too.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml [hunk 8]
```diff
@@ -25,7 +25,14 @@
     <label asp-for="Authority">@T["Authority"]</label>
     <input asp-for="Authority" class="form-control" />
     <span asp-validation-for="Authority"></span>
-    <span class="hint">@T["The address of the remote OpenID Connect server. Using a HTTPS address is strongly recommended."]</span>
+    <span class="hint">@T["The address of the remote OpenID Connect server. An HTTPS address is required."]</span>
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Views/OpenIdValidationSettings.Edit.cshtml:28]: Revert this as we need to keep supporting HTTP.

**Review by kevinchalet — APPROVED** [review]
I have a few nits but it looks good 👍🏻

**Comment by aderderian** [comment]
@kevinchalet  do I need to alert someone else now to approve it or run the workflows?

**Comment by kevinchalet** [comment]
@aderderian I just unlocked the build 😃

**Comment by kevinchalet** [comment]
PR merged, thanks for your contribution @aderderian.

I see you removed the URI scheme check and the ability to use a relative URI for the metadata address. Can you please open a follow-up ticket to ensure we track that?

Thanks.

**Comment by aderderian** [comment]
@kevinchalet  can you point me to where I should open the ticket? Thx!

**Comment by Skrypt** [comment]
@kevinchalet Did you mean opening a new issue in this repository to make changes to the OpenId integration settings?

**Pull request merged**